File: v4/rust/iprange-livedb/src/publication/replacement.rs
//! Stable ownership of the destination replaced by one publication.

use std::fs::File;

use crate::cancellation::CancellationToken;
use crate::live_lock::{self, Mode};
use crate::live_sidecar::MAIN_LIFETIME_LOCK;

use super::namespace::{
    regular_identity, regular_identity_any_link, regular_link_count, Destination, Identity, Name,
    NamespaceError,
};
use super::output::{self, PreparedOutput};
use super::reservation::Policy;

#[derive(Debug)]
pub(crate) struct PreviousMain {
    pub(crate) file: File,
    pub(crate) identity: Identity,
    pub(crate) byte_length: u64,
    pub(crate) sha512: [u8; 64],
}

#[derive(Debug)]
pub(crate) enum Error {
    Namespace(NamespaceError),
    Sdk(crate::error::Error),
    Output(output::Error),
    SameIdentity,
    ContentChanged,
}

#[derive(Debug)]
pub(crate) struct Failure {
    pub(crate) output: PreparedOutput,
    pub(crate) cause: Error,
}

pub(crate) fn bind(
    output: PreparedOutput,
    cancellation: &CancellationToken,
) -> Result<PreparedOutput, Box<Failure>> {
    bind_with(output, Policy::ReplaceExisting, cancellation)
}

pub(crate) fn bind_no_rollback(
    output: PreparedOutput,
    cancellation: &CancellationToken,
) -> Result<PreparedOutput, Box<Failure>> {
    bind_with(output, Policy::ReplaceExistingNoRollback, cancellation)
}

fn bind_with(
    mut output: PreparedOutput,
    policy: Policy,
    cancellation: &CancellationToken,
) -> Result<PreparedOutput, Box<Failure>> {
    debug_assert!(policy.is_replacement());
    match open(&output, cancellation) {
        Ok(previous) => {
            output.policy = policy;
            output.previous = Some(previous);
            Ok(output)
        }
        Err(cause) => Err(Box::new(Failure { output, cause })),
    }
}

fn open(output: &PreparedOutput, cancellation: &CancellationToken) -> Result<PreviousMain, Error> {
    cancellation.check().map_err(Error::Sdk)?;
    let destination = output.attempt.destination();
    destination
        .directory()
        .require_absent(destination.coordination())?;
    let regular = destination
        .directory()
        .open_regular(destination.main(), true)?
        .ok_or(NamespaceError::Missing)?;
    if regular.identity == output.attempt.identity() {
        return Err(Error::SameIdentity);
    }
    live_lock::lock_cancellable(
        &regular.file,
        MAIN_LIFETIME_LOCK,
        Mode::Exclusive,
        cancellation,
    )
    .map_err(Error::Sdk)?;
    verify_canonical(destination, &regular.file, regular.identity, None)?;
    super::namespace::sync_file(&regular.file).map_err(crate::error::Error::from)?;
    cancellation.check().map_err(Error::Sdk)?;
    let byte_length = regular
        .file
        .metadata()
        .map_err(crate::error::Error::from)?
        .len();
    let sha512 = output::digest_cancellable(&regular.file, byte_length, cancellation)
        .map_err(Error::Output)?;
    verify_canonical(
        destination,
        &regular.file,
        regular.identity,
        Some(byte_length),
    )?;
    Ok(PreviousMain {
        file: regular.file,
        identity: regular.identity,
        byte_length,
        sha512,
    })
}

impl PreviousMain {
    pub(crate) fn verify_canonical_namespace(
        &self,
        destination: &Destination,
    ) -> Result<(), NamespaceError> {
        verify_canonical(
            destination,
            &self.file,
            self.identity,
            Some(self.byte_length),
        )
    }

    pub(crate) fn verify_private_or_retired(
        &self,
        destination: &Destination,
        private_name: &Name,
    ) -> Result<(), NamespaceError> {
        destination.directory().verify()?;
        let identity = regular_identity_any_link(&self.file, destination.directory().identity())?;
        if identity != self.identity
            || self.file.metadata().map_err(NamespaceError::Io)?.len() != self.byte_length
        {
            return Err(NamespaceError::IdentityChanged);
        }
        match regular_link_count(&self.file)? {
            1 => destination
                .directory()
                .verify_name(private_name, self.identity),
            0 => destination.directory().require_absent(private_name),
            links => Err(NamespaceError::LinkCount(links)),
        }
    }

    pub(crate) fn verify_retired(
        &self,
        destination: &Destination,
        private_name: &Name,
    ) -> Result<(), NamespaceError> {
        destination.directory().verify()?;
        destination.directory().require_absent(private_name)?;
        let identity = regular_identity_any_link(&self.file, destination.directory().identity())?;
        if identity != self.identity {
            return Err(NamespaceError::IdentityChanged);
        }
        let metadata = self.file.metadata().map_err(NamespaceError::Io)?;
        if metadata.len() != self.byte_length {
            return Err(NamespaceError::IdentityChanged);
        }
        let links = regular_link_count(&self.file)?;
        if links != 0 {
            return Err(NamespaceError::LinkCount(links));
        }
        Ok(())
    }

    pub(crate) fn verify_content(
        &self,
        destination: &Destination,
        cancellation: Option<&CancellationToken>,
    ) -> Result<(), Error> {
        self.verify_canonical_namespace(destination)?;
        let digest = match cancellation {
            Some(cancellation) => {
                output::digest_cancellable(&self.file, self.byte_length, cancellation)
            }
            None => output::digest(&self.file, self.byte_length),
        }
        .map_err(Error::Output)?;
        self.verify_canonical_namespace(destination)?;
        if digest != self.sha512 {
            return Err(Error::ContentChanged);
        }
        Ok(())
    }
}

fn verify_canonical(
    destination: &Destination,
    file: &File,
    expected: Identity,
    expected_length: Option<u64>,
) -> Result<(), NamespaceError> {
    destination.directory().verify()?;
    destination
        .directory()
        .verify_name(destination.main(), expected)?;
    let identity = regular_identity(file, destination.directory().identity())?;
    if identity != expected {
        return Err(NamespaceError::IdentityChanged);
    }
    if expected_length.is_some_and(|length| {
        file.metadata()
            .map(|metadata| metadata.len() != length)
            .unwrap_or(true)
    }) {
        return Err(NamespaceError::IdentityChanged);
    }
    Ok(())
}

impl From<NamespaceError> for Error {
    fn from(value: NamespaceError) -> Self {
        Self::Namespace(value)
    }
}

impl From<crate::error::Error> for Error {
    fn from(value: crate::error::Error) -> Self {
        Self::Sdk(value)
    }
}

File: v4/go/internal/publication/replacement.go
// Stable ownership of the destination replaced by one publication
// (Rust publication/replacement.rs). The replacement bind opens the
// canonical main, proves it is not the attempt itself, locks it for
// the artifact lifetime, digests it, and re-proves it before and after
// the digest; the previous main is then retired by the main-file
// machine under the same lock.

package publication

import (
	"os"

	"github.com/firehol/iprange/v4/go/internal/format"
	"github.com/firehol/iprange/v4/go/internal/live"
	"github.com/firehol/iprange/v4/go/internal/mapping"
)

// previousMain is the replaced canonical main (Rust PreviousMain).
type previousMain struct {
	file       *os.File
	mapping    *mapping.Mapping
	identity   live.FileIdentity
	byteLength uint64
	sha512     [64]byte
}

// Close releases the previous-main mapping and descriptor (Rust drop
// of PreviousMain).
func (p *previousMain) Close() error {
	var first error
	if p.mapping != nil {
		if err := p.mapping.Close(); err != nil && first == nil {
			first = err
		}
	}
	if p.file != nil {
		if err := p.file.Close(); err != nil && first == nil {
			first = err
		}
	}
	return first
}

// replacementFailure is one replacement bind failure carrying the
// still-owned prepared output (Rust Failure<PreparedOutput> in
// replacement.rs).
type replacementFailure struct {
	output *preparedOutput
	cause  error
}

func (f *replacementFailure) Error() string { return f.cause.Error() }
func (f *replacementFailure) Unwrap() error { return f.cause }

// bindPrevious opens the replaced canonical main and attaches it to
// the prepared output under the replace-existing policy (Rust
// replacement::bind).
func bindPrevious(output *preparedOutput, check func() error) (*preparedOutput, *replacementFailure) {
	return bindPreviousWith(output, reservationPolicyReplaceExisting, check)
}

// bindPreviousNoRollback attaches the replaced canonical main under
// the replace-existing-no-rollback policy (Rust
// replacement::bind_no_rollback).
func bindPreviousNoRollback(output *preparedOutput, check func() error) (*preparedOutput, *replacementFailure) {
	return bindPreviousWith(output, reservationPolicyReplaceExistingNoRollback, check)
}

func bindPreviousWith(output *preparedOutput, policy reservationPolicy, check func() error) (*preparedOutput, *replacementFailure) {
	// rust: debug_assert!(policy.is_replacement())
	previous, err := openPrevious(output, check)
	if err != nil {
		return nil, &replacementFailure{output: output, cause: err}
	}
	output.policy = policy
	output.previous = previous
	return output, nil
}

// openPrevious opens and proves the replaced canonical main (Rust
// replacement.rs open).
func openPrevious(output *preparedOutput, check func() error) (*previousMain, error) {
	if err := live.Checkpoint(check); err != nil {
		return nil, err
	}
	destination := output.attempt.destination
	if err := destination.directory().RequireAbsent(destination.coordinationName()); err != nil {
		return nil, err
	}
	regular, err := destination.directory().OpenRegular(destination.mainName(), true)
	if err != nil {
		return nil, err
	}
	if regular == nil {
		return nil, &live.NamespaceError{Kind: live.NamespaceMissing}
	}
	if regular.Identity == output.attempt.identity {
		regular.File.Close()
		return nil, sameIdentityError()
	}
	if err := live.LockFileCancellable(regular.File, live.MainLifetimeOffset, live.LockExclusive, check); err != nil {
		regular.File.Close()
		return nil, err
	}
	if err := verifyCanonicalDestination(destination, regular.File, regular.Identity, 0, false); err != nil {
		regular.File.Close()
		return nil, err
	}
	if err := live.SyncFile(regular.File); err != nil {
		regular.File.Close()
		return nil, err
	}
	if err := live.Checkpoint(check); err != nil {
		regular.File.Close()
		return nil, err
	}
	st, err := regular.File.Stat()
	if err != nil {
		regular.File.Close()
		return nil, &format.Error{Code: format.CodeIO, Detail: "stat: " + err.Error()}
	}
	byteLength := uint64(st.Size())
	mapped, err := mapping.MapFile(regular.File, byteLength, false)
	if err != nil {
		regular.File.Close()
		return nil, err
	}
	sha512, err := digestCancellable(mapped, byteLength, check)
	if err != nil {
		mapped.Close()
		regular.File.Close()
		return nil, err
	}
	if err := verifyCanonicalDestination(destination, regular.File, regular.Identity, byteLength, true); err != nil {
		mapped.Close()
		regular.File.Close()
		return nil, err
	}
	return &previousMain{
		file:       regular.File,
		mapping:    mapped,
		identity:   regular.Identity,
		byteLength: byteLength,
		sha512:     sha512,
	}, nil
}

// sameIdentityError is the fixed same-identity class (Rust
// replacement::Error::SameIdentity, mapped to Problem::replacement by
// problem.go).
func sameIdentityError() error {
	return &format.Error{Code: format.CodeConflict, Detail: "replacement source and destination identities match"}
}

// contentChangedError is the fixed content-changed class (Rust
// replacement::Error::ContentChanged, mapped to Problem::replacement
// by problem.go).
func contentChangedError() error {
	return &format.Error{Code: format.CodeConflict, Detail: "replacement destination content changed"}
}

// verifyCanonicalNamespace proves the previous main is still the
// canonical main of the destination (Rust
// PreviousMain::verify_canonical_namespace).
func (p *previousMain) verifyCanonicalNamespace(destination *destination) error {
	return verifyCanonicalDestination(destination, p.file, p.identity, p.byteLength, true)
}

// verifyPrivateOrRetired proves the previous main is now at the
// private output name (one link) or fully unlinked (zero links) after
// the main rename (Rust PreviousMain::verify_private_or_retired).
func (p *previousMain) verifyPrivateOrRetired(destination *destination, privateName string) error {
	if err := destination.directory().Verify(); err != nil {
		return err
	}
	identity, err := live.RegularIdentityAnyLink(p.file, destination.directory().Identity())
	if err != nil {
		return err
	}
	size, err := fstatSize(p.file)
	if err != nil {
		return &live.NamespaceError{Kind: live.NamespaceIo, Op: "inspect retained file", Err: err}
	}
	if identity != p.identity || size != p.byteLength {
		return &live.NamespaceError{Kind: live.NamespaceIdentityChanged}
	}
	links, err := live.RegularLinkCount(p.file)
	if err != nil {
		return err
	}
	switch links {
	case 1:
		return destination.directory().VerifyName(privateName, p.identity)
	case 0:
		return destination.directory().RequireAbsent(privateName)
	default:
		return &live.NamespaceError{Kind: live.NamespaceLinkCount, Links: links}
	}
}

// verifyRetired proves the previous main is fully unlinked and gone
// from the destination (Rust PreviousMain::verify_retired).
func (p *previousMain) verifyRetired(destination *destination, privateName string) error {
	if err := destination.directory().Verify(); err != nil {
		return err
	}
	if err := destination.directory().RequireAbsent(privateName); err != nil {
		return err
	}
	identity, err := live.RegularIdentityAnyLink(p.file, destination.directory().Identity())
	if err != nil {
		return err
	}
	if identity != p.identity {
		return &live.NamespaceError{Kind: live.NamespaceIdentityChanged}
	}
	size, err := fstatSize(p.file)
	if err != nil {
		return &live.NamespaceError{Kind: live.NamespaceIo, Op: "inspect retained file", Err: err}
	}
	if size != p.byteLength {
		return &live.NamespaceError{Kind: live.NamespaceIdentityChanged}
	}
	links, err := live.RegularLinkCount(p.file)
	if err != nil {
		return err
	}
	if links != 0 {
		return &live.NamespaceError{Kind: live.NamespaceLinkCount, Links: links}
	}
	return nil
}

// verifyContent digests the retained previous main and compares it to
// the bind-time digest (Rust PreviousMain::verify_content). The
// check function is optional (Rust Option<CancellationToken>).
func (p *previousMain) verifyContent(destination *destination, check func() error) error {
	if err := p.verifyCanonicalNamespace(destination); err != nil {
		return err
	}
	var (
		digested [64]byte
		err      error
	)
	if check != nil {
		digested, err = digestCancellable(p.mapping, p.byteLength, check)
	} else {
		digested, err = digest(p.mapping, p.byteLength)
	}
	if err != nil {
		return err
	}
	if err := p.verifyCanonicalNamespace(destination); err != nil {
		return err
	}
	if digested != p.sha512 {
		return contentChangedError()
	}
	return nil
}

// verifyCanonicalDestination proves one open file is the canonical
// main of the destination with the expected identity and optionally
// the expected length (Rust replacement.rs verify_canonical).
func verifyCanonicalDestination(destination *destination, file *os.File, expected live.FileIdentity, expectedLength uint64, requireLength bool) error {
	if err := destination.directory().Verify(); err != nil {
		return err
	}
	if err := destination.directory().VerifyName(destination.mainName(), expected); err != nil {
		return err
	}
	identity, err := live.RegularIdentity(file, destination.directory().Identity())
	if err != nil {
		return err
	}
	if identity != expected {
		return &live.NamespaceError{Kind: live.NamespaceIdentityChanged}
	}
	if requireLength {
		st, err := file.Stat()
		if err != nil {
			return &live.NamespaceError{Kind: live.NamespaceIdentityChanged}
		}
		if uint64(st.Size()) != expectedLength {
			return &live.NamespaceError{Kind: live.NamespaceIdentityChanged}
		}
	}
	return nil
}
